lib/IRGen/IRGenDebugInfo.cpp

```diff
@@ -798,9 +798,28 @@ class IRGenDebugInfoImpl : public IRGenDebugInfo {
     std::string Result = Mangler.mangleTypeForDebugger(
         Ty, nullptr);
 
+    // Returns true iff the given type comes from an implicit declaration
+    // generated by the differentiation transform: a linear map struct or
+    // branching trace enum.
+    auto isAutoDiffGeneratedDeclarationType = [](Type type) -> bool {
+      // Peer through metatype types.
+      type = type->getMetatypeInstanceType();
+      // Check for linear map structs.
+      if (auto *structDecl = type->getStructOrBoundGenericStruct())
+        return structDecl->isImplicit() &&
+               structDecl->getNameStr().contains("__PB__");
+      // Check for branching trace enums.
+      if (auto *enumDecl = type->getEnumOrBoundGenericEnum())
+        return enumDecl->isImplicit() &&
+               enumDecl->getNameStr().contains("__Pred__");
+      return false;
+    };
+
     if (!Opts.DisableRoundTripDebugTypes
         // FIXME: implement type reconstruction for opaque types
-        && !Ty->hasOpaqueArchetype()) {
+        && !Ty->hasOpaqueArchetype()
+        // FIXME(TF-1232): Add type reconstruction for AutoDiff-generated decls.
+        && !isAutoDiffGeneratedDeclarationType(Ty)) {
```

Comment:
@adrian-prantl: cc'ing you since you're a debug info code owner! I hope it's okay to add a special case for skipping round-trip debug type verification, since other special cases exist.

I tried to support type reconstruction for AutoDiff-generated declarations (TF-1232), but it seems difficult due to poor name lookup support for late generated declarations (SILOptimizer) after name lookup caches have been populated.

Reply:
What's the plan for supporting it? In my experience grafting on debug info support as an afterthought is going to be even harder, because changing the design after the fact always has more friction. Can you explain what kinds of features are missing from SILOptimizer and other components that are needed to make this work?

Reply:
Thanks for responding!

> Can you explain what kinds of features are missing from SILOptimizer and other components that are needed to make this work?

Some context: the differentiation transform in SILOptimizer generates some implicit `FileUnit`-level struct and enum declarations associated with `SILFunction`s ([here](https://github.com/apple/swift/blob/905341321ba8f981efa609f9d358d45a48002cb8/include/swift/SILOptimizer/Utils/Differentiation/LinearMapInfo.h#L112-L125)). AFAIK, it's unprecedented to create declarations and add them to `FileUnit`s so late in the compiler pipeline.

---

I believe the main issue is: these implicit declarations aren't being correctly added to `FileUnit`s so that they can be found via name lookup.

Currently, name lookup (`ModuleDecl::lookupMember` in `ASTBuilder::findTypeDecl`) fails during round-trip debug type verification ([from test](https://github.com/apple/swift/blob/fccfa29240d1f151d2da0d2bf2c7eff067a1f4c5/test/AutoDiff/compiler_crashers/tf1232-autodiff-generated-declaration-mangling.swift)):
```
Failed to reconstruct type for $s4main42_AD__$s4main3fooyS2fF_bb0__PB__src_0_wrt_0VmD
Original type:
(metatype_type
  (struct_type decl=main.(file)._AD__$s4main3fooyS2fF_bb0__PB__src_0_wrt_0))
```

---

Do you happen to have any context on adding declarations to `FileUnit`s for name lookup?

I explored a fix in TF-1232, but it involved calling `ModuleDecl::clearLookupCache`, which is marked with "do not add more uses". I don't know whether adding another use is acceptable.

I'll continue investigating a robust fix.

Reply:
I think instead of trying to update name lookup to handle this, it would be better to add the synthesized declarations to a special list somewhere, and teach type reconstruction to go look at that list.